genesyscloud/resource_genesyscloud_outbound_dnclist.go

```diff
@@ -0,0 +1,376 @@
+		}
+
+		cc := consistency_checker.NewConsistencyCheck(ctx, d, meta, resourceOutboundDncList())
+
+		if sdkDncList.Name != nil {
+			_ = d.Set("name", *sdkDncList.Name)
+		}
+		if sdkDncList.ContactMethod != nil {
+			_ = d.Set("contact_method", *sdkDncList.ContactMethod)
+		}
+		if sdkDncList.LoginId != nil {
+			_ = d.Set("login_id", *sdkDncList.LoginId)
+		}
+		if sdkDncList.CampaignId != nil {
+			_ = d.Set("campaign_id", *sdkDncList.CampaignId)
+		}
+		if sdkDncList.DncCodes != nil {
+			var dncCodes []string
+			schemaCodes := interfaceListToStrings(d.Get("dnc_codes").([]interface{}))
+			if len(*sdkDncList.DncCodes) == len(schemaCodes) {
+				// Rearrange to the original ordering
+				for _, schemaCode := range schemaCodes {
+					for _, sdkCode := range *sdkDncList.DncCodes {
+						if schemaCode == sdkCode {
+							dncCodes = append(dncCodes, sdkCode)
+						}
+					}
+				}
+				if len(dncCodes) != len(*sdkDncList.DncCodes) {
+					dncCodes = nil
+				}
+			}
+			if len(dncCodes) == 0 {
+				for _, v := range *sdkDncList.DncCodes {
+					dncCodes = append(dncCodes, v)
+				}
+			}
+			_ = d.Set("dnc_codes", dncCodes)
+		}
+		if sdkDncList.LicenseId != nil {
+			_ = d.Set("license_id", *sdkDncList.LicenseId)
+		}
+		if sdkDncList.Division != nil && sdkDncList.Division.Id != nil {
+			_ = d.Set("division_id", *sdkDncList.Division.Id)
+		}
+		log.Printf("Read Outbound DNC list %s %s", d.Id(), *sdkDncList.Name)
+		return cc.CheckState()
+	})
+}
+
+func deleteOutboundDncList(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sdkConfig := meta.(*providerMeta).ClientConfig
+	outboundApi := platformclientv2.NewOutboundApiWithConfig(sdkConfig)
+
+	diagErr := retryWhen(isStatus400, func() (*platformclientv2.APIResponse, diag.Diagnostics) {
+		log.Printf("Deleting Outbound DNC list")
+		resp, err := outboundApi.DeleteOutboundDnclist(d.Id())
+		if err != nil {
+			return resp, diag.Errorf("Failed to delete Outbound DNC list: %s", err)
+		}
+		return resp, nil
+	})
+	if diagErr != nil {
+		return diagErr
+	}
+
+	return withRetries(ctx, 30*time.Second, func() *resource.RetryError {
+		_, resp, err := outboundApi.GetOutboundDnclist(d.Id(), false, false)
+		if err != nil {
+			if isStatus404(resp) {
+				// Outbound DNC list deleted
+				log.Printf("Deleted Outbound DNC list %s", d.Id())
+				return nil
+			}
+			return resource.NonRetryableError(fmt.Errorf("error deleting Outbound DNC list %s: %s", d.Id(), err))
+		}
+
+		return resource.RetryableError(fmt.Errorf("Outbound DNC list %s still exists", d.Id()))
+	})
+}
+
+func uploadPhoneEntriesToDncList(api *platformclientv2.OutboundApi, dncList *platformclientv2.Dnclist, entry interface{}) (*platformclientv2.APIResponse, diag.Diagnostics) {
+	var phoneNumbers []string
+	if entryMap, ok := entry.(map[string]interface{}); ok && len(entryMap) > 0 {
+		if phoneNumbersList := entryMap["phone_numbers"].([]interface{}); phoneNumbersList != nil {
```

Comment:
Are we testing this bit with the phone_numbers? It would be great to get some coverage on it

Reply:
The best I could do was check that _n_ fields were uploaded using **GET /api/v2/outbound/dnclists/divisionviews/{dncListId}** because that endpoint lets you know how many rows are in the list. If I wanted to verify the actual values and dates, I'd be back into exporting the csv data, opening the browser, etc. At this point I've only done that through manual acceptance testing. 
If you think it's better to leave this functionality out, that's no prob.

Reply:
Na this looks good enough. The tests seem to cover it well.